### PR TITLE
Add a ServiceAccounts dropdown component

### PR DESCRIPTION
**Pull request by ncskier**
# Changes

Adding a ServiceAccounts dropdown component that can be reused many places in our future UI.
This PR adds a new `serviceaccounts` endpoint to the dashboard with accompanying documentation in `DEVELOPMENT.md`. Backend and frontend test files are included as well as a storybook file.

I also updated the namespaces dropdown to have a default label "Select Namespace" to match the service accounts dropdown default label "Select ServiceAccount". The label will appear when nothing is selected, and as a prop it can be overwritten.

Thanks @jessm12 for the initial work on this component.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/dashboard/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: src/actions/serviceAccounts.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { getServiceAccounts } from '../api';
+import { getSelectedNamespace } from '../reducers';
+
+export function fetchServiceAccountsSuccess(data, namespace) {
```
Comment (AlanGreene) [inline src/actions/serviceAccounts.js]: We've made some changes recently to the action creators and reducers. These changes should also be made for the new serviceaccounts functionality, both for consistency, and to ensure correct behaviour.

- changes to support loading all namespaces https://github.com/tektoncd/dashboard/pull/170
  - remove `namespace` from the success action creators
  - update reducers to read namespace value from the resource metadata
- allow overriding the selected namespace https://github.com/tektoncd/dashboard/pull/164
  - update 'fetch' action creators to accept namespace as a parameter (for use by extensions)
- expose action creators to extensions (where appropriate)
  - https://github.com/tektoncd/dashboard/blob/master/src/containers/Extension/actions.js

Comment (ncskier) [reply src/actions/serviceAccounts.js]: Thanks for the feedback Alan. I'll look over these and make the changes.

**Comment by ncskier** [comment]
@AlanGreene thanks for the review. I've updated the action creators and reducers to support loading all namespaces. I also added the fetch serviceaccounts action creator to extensions.

Let me know if there's anything else I should change 🙂

File: src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      this.props.fetchServiceAccounts();
+    }
+  }
+
+  render() {
+    const { loading, ...dropdownProps } = this.props;
+    if (loading) {
+      return <DropdownSkeleton {...dropdownProps} />;
+    }
+    return <Dropdown {...dropdownProps} />;
```
Comment (AlanGreene) [inline src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.js]: The Dropdown component truncates long options but does not display a tooltip. To allow users to correctly identify the options we made some additions to the NamespacesDropdown to add a `title` attribute on the options. That could should be refactored for reuse here.

https://github.com/tektoncd/dashboard/blob/fc9259d40e958dfe2d3f6bd8307c58a1aafaed10/src/containers/NamespacesDropdown/NamespacesDropdown.js#L22

see the `itemToElement` and `itemToString` props passed to the Dropdown component

Comment (ncskier) [reply src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.js]: Ok, will do

**Comment by ncskier** [comment]
Thanks Alan, I added the `title` attribute just like with the `NamespacesDropdown`

**Comment by ncskier** [comment]
I updated the tests, so they cover the ServiceAccount Dropdown's `componentDidUpdate` function. @AlanGreene please let me know if there's anything else I can do for this PR 🙂

File: src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.test.js [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+  green: '',
+  blue: ''
+};
+
+const initialTextRegExp = new RegExp('select serviceaccount', 'i');
+
+const middleware = [thunk];
+const mockStore = configureStore(middleware);
+
+jest
```
Comment (AlanGreene) [inline src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.test.js]: `spyOn` and other setup should be called from inside an `it` to allow for proper cleanup between tests

if there's common code you want to execute in all of the tests you can use a `beforeEach`

File: src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+import { Dropdown, DropdownSkeleton } from 'carbon-components-react';
+
+import {
+  getServiceAccounts,
+  isFetchingServiceAccounts,
+  getSelectedNamespace
+} from '../../reducers';
+import { fetchServiceAccounts } from '../../actions/serviceAccounts';
+
+const itemToElement = ({ text }) => {
```
Comment (AlanGreene) [inline src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.js]: I could have been clearer in my previous comment, when I mentioned reuse I meant that the core of the namespaces dropdown should be refactored to extract the common pieces as a presentational component with the native tooltips for truncated text etc. Then both the namespaces dropdown and this new service accounts dropdown would build on top of that.

Ideally:
- Dropdown (the new presentational component) would
  - provide the loading state using DropdownSkeleton
  - provide these `itemToElement` and `itemToString` implementations by default to the Carbon component
  - automatically transform the `items` prop using `itemStringToObject`

- The `render` function of ServiceAccountsDropdown could look something like this:

```js
  render() {
    return (
      <Dropdown
        {...this.props}
        selectedItem={this.state.selectedItem}
        onChange={this.onChange}
      />
    );
  }
```
where `Dropdown` is the new component I described above

Comment (ncskier) [reply src/containers/ServiceAccountsDropdown/ServiceAccountsDropdown.js]: Ah, that sounds like the right way to do it 😅 I'll make the change.

**Comment by ncskier** [comment]
Thank you for the feedback @AlanGreene. I have addressed your two change requests, and the common tooltip dropdown code is now in a presentational `TooltipDropdown` component. Let me know what you think 🙂

**Pull request merged**